File: go/forge/forge.go
// SPDX-License-Identifier: MIT

// Package forge implements aether-forge: a mesh-native package cache proxy
// (Phase-2 extension). The first internet pull of a package is cached as Aether
// content; subsequent pulls by anyone in the mesh are served locally at mesh
// speeds. Port of the C# reference (AetherNet.Forge). Ecosystems: npm, pip,
// cargo, go, nuget, git.
package forge

import (
	"context"
	"sort"
	"sync"
	"time"
)

// Entry is the metadata record for one cached package artifact. Package IDs use
// a namespaced "ecosystem:name@version" format (e.g. "npm:react@18.2.0").
type Entry struct {
	ContentHash   string
	PackageID     string
	FetchedAt     time.Time
	SizeBytes     int64
	DownloadCount int
}

// Stats are aggregate statistics for the local Forge cache.
type Stats struct {
	TotalBytesSaved  int64
	TotalPeersServed int
	CatalogueSize    int
	TopPackages      []Entry // most-downloaded first, up to 10
}

// Service is the mesh-native package cache.
type Service interface {
	// Query looks up a cached entry by package ID; returns nil if not cached.
	Query(ctx context.Context, packageID string) (*Entry, error)
	// Cache stores a new artifact (idempotent — first write wins).
	Cache(ctx context.Context, packageID, contentHash string, sizeBytes int64) (*Entry, error)
	// Fetch increments the download counter and returns the entry, or nil if not cached.
	Fetch(ctx context.Context, packageID string) (*Entry, error)
	// GetStats returns current aggregate cache statistics.
	GetStats(ctx context.Context) (Stats, error)
}

// InMemoryService is an in-memory Service for testing / single-node use; state
// is lost on restart.
type InMemoryService struct {
	mu    sync.Mutex
	store map[string]*Entry // key = packageID

	// OnNewEntryAnnounced fires when a new artifact is added via Cache.
	OnNewEntryAnnounced func(*Entry)
}

// NewInMemoryService constructs an empty in-memory forge service.
func NewInMemoryService() *InMemoryService {
	return &InMemoryService{store: make(map[string]*Entry)}
}

// Query implements Service.
func (s *InMemoryService) Query(ctx context.Context, packageID string) (*Entry, error) {
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	e, ok := s.store[packageID]
	if !ok {
		return nil, nil
	}
	cp := *e
	return &cp, nil
}

// Cache implements Service (idempotent: an existing packageID is returned unchanged).
func (s *InMemoryService) Cache(ctx context.Context, packageID, contentHash string, sizeBytes int64) (*Entry, error) {
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	s.mu.Lock()
	e, ok := s.store[packageID]
	isNew := false
	if !ok {
		e = &Entry{
			PackageID:     packageID,
			ContentHash:   contentHash,
			SizeBytes:     sizeBytes,
			FetchedAt:     time.Now().UTC(),
			DownloadCount: 0,
		}
		s.store[packageID] = e
		isNew = true
	}
	cp := *e
	s.mu.Unlock()

	if isNew && s.OnNewEntryAnnounced != nil {
		s.OnNewEntryAnnounced(&cp)
	}
	return &cp, nil
}

// Fetch implements Service.
func (s *InMemoryService) Fetch(ctx context.Context, packageID string) (*Entry, error) {
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	e, ok := s.store[packageID]
	if !ok {
		return nil, nil
	}
	e.DownloadCount++
	cp := *e
	return &cp, nil
}

// GetStats implements Service.
func (s *InMemoryService) GetStats(ctx context.Context) (Stats, error) {
	if err := ctx.Err(); err != nil {
		return Stats{}, err
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	entries := make([]Entry, 0, len(s.store))
	var totalBytesSaved int64
	for _, e := range s.store {
		entries = append(entries, *e)
		totalBytesSaved += int64(e.DownloadCount) * e.SizeBytes
	}
	sort.SliceStable(entries, func(i, j int) bool {
		return entries[i].DownloadCount > entries[j].DownloadCount
	})
	top := entries
	if len(top) > 10 {
		top = top[:10]
	}
	return Stats{
		TotalBytesSaved:  totalBytesSaved,
		TotalPeersServed: 0, // no peer tracking in the in-memory implementation
		CatalogueSize:    len(s.store),
		TopPackages:      top,
	}, nil
}

File: rust/src/forge.rs
// SPDX-License-Identifier: MIT
//! aether-forge: a mesh-native package cache proxy (Phase-2 extension).
//!
//! The first internet pull of a package is cached as Aether content; subsequent
//! pulls by anyone in the mesh are served locally at mesh speeds. Port of the C#
//! reference (AetherNet.Forge). Ecosystems: npm, pip, cargo, go, nuget, git.

use std::collections::HashMap;
use std::time::{SystemTime, UNIX_EPOCH};

fn now_secs() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs())
        .unwrap_or(0)
}

/// Metadata record for one cached package artifact. Package IDs use a namespaced
/// "ecosystem:name@version" format (e.g. "npm:react@18.2.0").
#[derive(Clone, Debug)]
pub struct ForgeEntry {
    pub content_hash: String,
    pub package_id: String,
    /// Unix-epoch seconds when first fetched and cached.
    pub fetched_at_secs: u64,
    pub size_bytes: i64,
    pub download_count: i32,
}

/// Aggregate statistics for the local Forge cache.
#[derive(Clone, Debug, Default)]
pub struct ForgeStats {
    pub total_bytes_saved: i64,
    pub total_peers_served: i32,
    pub catalogue_size: i32,
    /// Most-downloaded first, up to 10.
    pub top_packages: Vec<ForgeEntry>,
}

type Callback = Box<dyn Fn(&ForgeEntry)>;

/// In-memory mesh-native package cache for testing / single-node use; state is
/// lost on drop.
#[derive(Default)]
pub struct InMemoryForgeService {
    store: HashMap<String, ForgeEntry>, // key = package_id
    pub on_new_entry_announced: Option<Callback>,
}

impl InMemoryForgeService {
    pub fn new() -> Self {
        Self::default()
    }

    /// Look up a cached entry by package ID; `None` if not cached.
    pub fn query(&self, package_id: &str) -> Option<ForgeEntry> {
        self.store.get(package_id).cloned()
    }

    /// Store a new artifact (idempotent — an existing package_id is returned unchanged).
    pub fn cache(&mut self, package_id: &str, content_hash: &str, size_bytes: i64) -> ForgeEntry {
        if let Some(existing) = self.store.get(package_id) {
            return existing.clone(); // first-write-wins
        }
        let entry = ForgeEntry {
            content_hash: content_hash.to_string(),
            package_id: package_id.to_string(),
            fetched_at_secs: now_secs(),
            size_bytes,
            download_count: 0,
        };
        self.store.insert(package_id.to_string(), entry.clone());
        if let Some(cb) = &self.on_new_entry_announced {
            cb(&entry);
        }
        entry
    }

    /// Increment the download counter and return the entry, or `None` if not cached.
    pub fn fetch(&mut self, package_id: &str) -> Option<ForgeEntry> {
        let entry = self.store.get_mut(package_id)?;
        entry.download_count += 1;
        Some(entry.clone())
    }

    /// Current aggregate cache statistics.
    pub fn get_stats(&self) -> ForgeStats {
        let mut entries: Vec<ForgeEntry> = self.store.values().cloned().collect();
        let total_bytes_saved: i64 = entries
            .iter()
            .map(|e| e.download_count as i64 * e.size_bytes)
            .sum();
        entries.sort_by(|a, b| b.download_count.cmp(&a.download_count));
        let top_packages: Vec<ForgeEntry> = entries.iter().take(10).cloned().collect();
        ForgeStats {
            total_bytes_saved,
            total_peers_served: 0,
            catalogue_size: self.store.len() as i32,
            top_packages,
        }
    }
}
